### PR TITLE
[web-console] Reconcile high-throughput stream parsing behavior

### PR DESCRIPTION
**Pull request by Karakatiza666**
Treat metrics stream values as simple numbers

Testing: manually make sure the metrics graphs, the change stream and logs behave as expected, in particular Change Stream is responsive when displaying a high-throughput stream.

The change removes a previous implementation for stream parsing, now Logs and Change Stream use the sam eload shedding mechanism; metrics parsing uses a simpler, off-the-shelf stream parsing without load shedding, and removes the unnecessary complication of using a BigNumber implementation to process the metrics values. The total code size and complexity is reduced.

### CODE REVIEW

File: js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte [hunk 1]
```diff
@@ -90,45 +87,39 @@
...
-          if (metricsAvailable && cancelStream) {
-            endMetricsStream()
-            if (pipelineName !== targetPipelineName) {
-              return
-            }
-            startMetricsStream(api, targetPipelineName)
-          }
+    // `JSONParser` from `@streamparser/json-whatwg` handles tokenizing and
+    // document framing. Metric values parse as JS numbers — `Number.MAX_SAFE_INTEGER`
+    // is well outside the range of realistic timestamps, record counts, and byte sizes.
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte]: why is the comment about SAFE_INTEGER relevant? I don't see it used anywhere.

File: js-packages/web-console/src/lib/functions/pipelines/changeStream.ts [hunk 2]
```diff
@@ -276,321 +243,267 @@ export const parseStream = <T>(
 }
 
 /**
+ * Decoder for newline-delimited JSON. Two granularities are at play here:
+ *
+ *   - **Network chunk:** whatever bytes one `reader.read()` call returns.
+ *     Sized by the network, unpredictable, can be very large.
+ *   - **Parser chunk:** a line-aligned slice the decoder hands to
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/functions/pipelines/changeStream.ts:250]: what happens if you have a very large string in a column?

File: js-packages/web-console/src/lib/functions/pipelines/changeStream.ts [hunk 3]
```diff
@@ -276,321 +243,267 @@ export const parseStream = <T>(
...
+ *     `parser.write` — bounded by `PARSER_CHUNK_TARGET_BYTES` so a single
+ *     synchronous parse can't monopolize the main thread.
+ *
+ * After each read, the accumulated buffer (prior leftover + the new network
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/functions/pipelines/changeStream.ts:254]: I recall reading these comments in a prior PR.
How is this code related? Is it a new implementation?

**Review by mihaibudiu — APPROVED** [review]
Please fix the spelling of the commit message; I changed the PR.

File: js-packages/web-console/src/lib/functions/pipelines/changeStream.ts [hunk 4]
```diff
@@ -276,321 +243,267 @@ export const parseStream = <T>(
...
- * skipped and discarded, since accepting them would let a runaway record exhaust memory.
+ * @param opts.bufferSize Bytes admitted per `bufferWindowMs`; default 1 MB.
+ * @param opts.bufferWindowMs Budget reset period; default 100ms. Independent
+ *   of the orchestrator's UI flush cadence.
+ * @param opts.onBytesSkipped Called when a parser chunk is dropped, or when
+ *   a runaway record (no newline within `MAX_LINE_SIZE` bytes) is purged.
  */
-function splitByNewlineWithOverflowShedding(
-  maxBufferedBytes: number,
+export const newlineTextDecoder = (opts?: {
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/functions/pipelines/changeStream.ts:413]: **Hard block: `newlineTextDecoder` has zero test coverage.**

`newlineJsonDecoder` is exercised by the existing `parseStream` describe block, but `newlineTextDecoder` — a brand-new ~100-line decoder with its own forward-scan chunking strategy (different from `newlineJsonDecoder`'s backward `lastIndexOf` approach) — has no tests at all. It's used by `TabLogs.svelte` for the log viewer.

At minimum, please add a `describe('newlineTextDecoder')` block covering:
1. Lines split correctly across chunk boundaries
2. `\r\n` preserved on each emitted line (the docstring promises this)
3. Load shedding fires `onBytesSkipped` when budget is exceeded
4. Runaway record (no newline within `MAX_LINE_SIZE`) is purged

Vitest is already wired — this is straightforward.

File: js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte [hunk 5]
```diff
@@ -90,45 +87,39 @@
...
+        /* fall through to restart logic */
       }
-    )
-    cancelStream = cancel
+      if (cancelled || !metricsAvailable || !cancelStream) return
+      endMetricsStream()
+      if (pipelineName === targetPipelineName) {
+        startMetricsStream(api, targetPipelineName)
+      }
+    })()
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte]: This PR introduces `StreamFormatDecoder` + `parseStream` as *the* way to consume byte streams, then immediately sidesteps it here by piping through `JSONParser` from `@streamparser/json-whatwg` with a hand-rolled read loop. That means metrics streaming has:

- No load shedding (the other two consumers have it)
- No shared error/cancel lifecycle
- A bespoke restart pattern instead of `onParseEnded`
- A silent `catch {}` at line 114 that swallows all errors, not just network failures

Consider writing a thin decoder (e.g. `nativeJsonDecoder`) that wraps `JSONParser` behind the `StreamFormatDecoder` interface and plugs into `parseStream` like the other two call sites. That way the abstraction this refactor introduces actually applies uniformly.

File: js-packages/web-console/src/lib/functions/pipelines/changeStream.spec.ts [hunk 6]
```diff
@@ -59,17 +60,19 @@ const runParseStream = <T>(
...
             values.push(v)
           }
         },
-        onBytesSkipped: (n) => skipped.push(n),
         onParseEnded: () => resolve({ values, skipped, parser })
       },
-      options
+      { flushIntervalMs: options?.flushIntervalMs }
     )
   })
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/functions/pipelines/changeStream.spec.ts:78]: The test helper `runParseStream` only exercises `newlineJsonDecoder`. There is no equivalent for `newlineTextDecoder`, which has a substantially different chunking algorithm (forward `indexOf` scan collecting line endings vs. backward `lastIndexOf` snap). The two decoders could diverge on edge cases (e.g. a parser chunk that lands exactly on a `\n`, or a `\r\n` split across network chunks) without any test catching it.

**Review by mythical-fred — CHANGES_REQUESTED** [review]
Two blockers, see inline. The refactor itself is well-designed — the decoder/orchestrator split is a good separation of concerns. But the new `newlineTextDecoder` ships with zero tests, and TabPerformance opts out of the abstraction this PR introduces.

File: js-packages/web-console/src/lib/functions/pipelines/changeStream.ts (hunk 2 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/functions/pipelines/changeStream.ts:250]: Both the custom load-shedding decoder and the streaming JSON parser handle this correctly - a singlevalue split across multiple raw byte stream chunks

File: js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte (hunk 1 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte]: It is not used directly, but the comment describes why using some BigNumber implementation is not necessary, and the precision of the built-in JS Number primitive is sufficient to process metrics

**Review by mythical-fred — APPROVED** [review]
LGTM. Both prior blockers resolved: `newlineTextDecoder` now has thorough test coverage (line splitting, cross-chunk reassembly, CRLF, load shedding, boundary cases, runaway records) and the commit typo is fixed.

Soft note carried: `TabPerformance.svelte` still bypasses the new `parseStream` + decoder abstraction and uses `JSONParser` directly. Not blocking — design choice for the metrics stream. The silent `catch {}` in that component is still there; consider logging it eventually.
